include/util/static_rtree.hpp

```diff
@@ -49,71 +49,150 @@ namespace osrm
+     *          and create K..O with bounding boxes surrounding the groups of
+     *          BRANCHING_FACTOR.  Again, O, the last entry, may have fewer than
+     *          BRANCHING_FACTOR entries.
+     * Step 3a- Repeat this process for each level, until you only create 1 TreeNode
+     *          to contain its children (in this case, W).
+     *
+     * As we create TreeNodes, we append them to the m_search_tree vector.
+     *
+     * After this part of the building process, m_search_tree will contain TreeNode
+     * objects in this order:
+     *
+     * ABCDEFGHIJ KLMNO PQR UV W
+     * 10         5     3   2  1  <- number of nodes in the level
+     *
+     * In order to make our math easy later on, we reverse the whole array,
+     * then reverse the nodes within each level:
+     *
+     *   Reversed:        W VU RQP ONMKL JIHGFEDCBA
+     *   Levels reversed: W UV PQR KLMNO ABCDEFGHIJ
+     *
+     * We also now have the following information:
+     *
+     *   level sizes = {1,2,3,5,10}
+     *
+     * and we can calculate the array position the nodes for each level
+     * start (based on the sum of the previous level sizes):
+     *
+     *   level starts = {0,1,3,6,11}
+     *
+     * Now, some basic math can be used to navigate around the tree.  See
+     * the body of the `child_indexes` function for the details.
+     *
+     ***********************************************/
     template <typename T> using Vector = ViewOrVector<T, Ownership>;
 
   public:
     using Rectangle = RectangleInt2D;
     using EdgeData = EdgeDataT;
     using CoordinateList = Vector<util::Coordinate>;
 
-    static_assert(LEAF_PAGE_SIZE >= sizeof(uint32_t) + sizeof(EdgeDataT), "page size is too small");
+    static_assert(LEAF_PAGE_SIZE >= sizeof(EdgeDataT), "page size is too small");
     static_assert(((LEAF_PAGE_SIZE - 1) & LEAF_PAGE_SIZE) == 0, "page size is not a power of 2");
-    static constexpr std::uint32_t LEAF_NODE_SIZE =
-        (LEAF_PAGE_SIZE - sizeof(uint32_t) - sizeof(Rectangle)) / sizeof(EdgeDataT);
+    static constexpr std::uint32_t LEAF_NODE_SIZE = (LEAF_PAGE_SIZE / sizeof(EdgeDataT));
 
     struct CandidateSegment
     {
         Coordinate fixed_projected_coordinate;
         EdgeDataT data;
     };
 
+    /**
+     * Represents a node position somewhere in our tree.  This is purely a navigation
+     * class used to find children of each node - the actual data for each node
+     * is in the m_search_tree vector of TreeNode objects.
+     */
     struct TreeIndex
     {
-        TreeIndex() : index(0), is_leaf(false) {}
-        TreeIndex(std::size_t index, bool is_leaf) : index(index), is_leaf(is_leaf) {}
-        std::uint32_t index : 31;
-        std::uint32_t is_leaf : 1;
+        TreeIndex() : level(0), offset(0) {}
+        TreeIndex(std::size_t level_, std::size_t offset_) : level(level_), offset(offset_) {}
+        std::size_t level;  // Which level of the tree is this node in
```

Comment:
Going smaller here would help, the size of `TreeIndex` is important for the performance of the queue: The smaller this struct is the less data will be shuffled around. Going `size_t` -> `uint32` should work for both. (maybe even 16 bits?)